torch/distributed/fsdp/flat_param.py

```diff
@@ -575,6 +586,404 @@ def shard_metadata(
+            or flat_param.grad.device != flat_param.device  # grad on CPU
+        ):
+            self._check_on_compute_device(self.flat_param)
+            grad_offloaded = flat_param.grad.device != self.device
+            p_assert(
+                not grad_offloaded or self._config.offload_params,
+                f"Expects the sharded gradient to be on {self.device} "
+                f"but got {flat_param.grad.device}",
+            )
+            prev_iter_synced_gradients = (
+                flat_param.grad.size()
+                == flat_param._local_shard.size()  # type: ignore[attr-defined]
+            )
+            if prev_iter_synced_gradients:
+                # TODO (awgu): Gradient accumulation outside `no_sync()`
+                # does not work with CPU offloading. The issue should be
+                # that, in the post-backward hook, we cannot do an addition
+                # between a CPU tensor (the existing sharded gradient) and
+                # a GPU tensor (the new sharded gradient).
+                if not grad_offloaded:
+                    flat_param._saved_grad_shard = flat_param.grad.data  # type: ignore[attr-defined]
+            else:
+                padded_unsharded_size = flat_param._full_param_padded.size()  # type: ignore[attr-defined]
+                p_assert(
+                    flat_param.grad.size() == padded_unsharded_size,
+                    "Expects `.grad` to be the unsharded gradient in "
+                    f"`no_sync()` with size {padded_unsharded_size} "
+                    f"but got size {flat_param.grad.size()}",
+                )
+            flat_param.grad = None
+
+    @contextlib.contextmanager
+    def to_cpu(self):
+        """
+        Moves the unpadded unsharded flattened parameter to CPU while in the
+        context and moves it back to the previous device upon exit. For now,
+        this assumes the ``FlatParameter`` is the unpadded unsharded flattened
+        parameter since (1) there is no reason to include the padding in the
+        copy and (2) there is no use case for the sharded flattened parameter.
+
+        Precondition: ``self.flat_param`` 's data is the unpadded unsharded
+            flattened parameter on the compute device, and the handle uses a
+            sharded strategy.
+        Postcondition: Same as the precondition.
+        """
+        self._check_sharded_strategy()
+        p_assert(
+            self.flat_param.size() == self.flat_param._unsharded_size,
+            f"Expects size {self.flat_param._unsharded_size} but got {self.flat_param.size()}",
+        )
+        self._check_on_compute_device(self.flat_param)
+        # Check that the unpadded unsharded flattened parameter is a view into
+        # the padded unsharded flattened parameter as expected
+        # TODO (awgu): For correctness, this check is not strictly needed.
+        # However, the user should not be doing parameter swapping with
+        # `self.flat_param` either.
```

Comment:
seems in most cases, unsharded_flat_param data is used in FSDP internally, and it should not be exposed to users at all so it has rare chances to be mutated by users. what users can access mostly is sharded_flat_param data?

also, wondering whether we can just relax the check as mentioned in your TODO as well, we can allow user to mutate the data just in case, instead of checking it is mutated or not, we always do 'self._get_padded_unsharded_flat_param() = self.flat_param.data' to make them point to the same storage if that matters

Reply:
Since currently `to_cpu()` is only meant to be used for `summon_full_params()`, which itself _should_ ensure that this assertion succeeds, I think keeping this check is fine. If in the future there are other applications of `to_cpu()`, we can relax the check then.